src/main/java/org/broadinstitute/hellbender/tools/spark/bwa/BwaSpark.java

```diff
@@ -0,0 +1,328 @@
+            return Iterators.transform(fromIterator, new com.google.common.base.Function<F, T>() {
+                @Nullable
+                @Override
+                public T apply(@Nullable F input) {
+                    return function.apply(input);
+                }
+            });
+        }
+        // use an executor service for the multi-threaded case
+        final ExecutorService executorService = Executors.newFixedThreadPool(numThreads);
+        final Queue<Future<T>> futures = new LinkedList<>();
+        return new AbstractIterator<T>() {
+            @Override
+            protected T computeNext() {
+                try {
+                    while (fromIterator.hasNext()) {
+                        if (futures.size() == numThreads) {
+                            return futures.remove().get();
+                        }
+                        F next = fromIterator.next();
+                        Future<T> future = executorService.submit(() -> function.apply(next));
+                        futures.add(future);
+                    }
+                    if (!futures.isEmpty()) {
+                        return futures.remove().get();
+                    }
+                    executorService.shutdown();
+                    return endOfData();
+                } catch (InterruptedException | ExecutionException e) {
+                    throw new GATKException("Problem running task", e);
+                }
+            }
+        };
+    }
+
+    private static void reportThreadTimings() {
+        long bwaTime = 0;
+        for (Stopwatch stopwatch : stopwatches) {
+            bwaTime += stopwatch.elapsed(TimeUnit.MILLISECONDS);
+        }
+        log.info("Time in bwa: " + (bwaTime / 1000) + "s");
+    }
+
+    // From CreateSequenceDictionary
+
+    public int NUM_SEQUENCES = Integer.MAX_VALUE;
+
+    SAMSequenceDictionary makeSequenceDictionary(final File referenceFile) {
+        return makeSequenceDictionary(referenceFile.toPath());
+    }
+
+    SAMSequenceDictionary makeSequenceDictionary(final Path referenceFile) {
+        final ReferenceSequenceFile refSeqFile =
+                ReferenceSequenceFileFactory.getReferenceSequenceFile(referenceFile, true);
+        ReferenceSequence refSeq;
+        final List<SAMSequenceRecord> ret = new ArrayList<>();
+        final Set<String> sequenceNames = new HashSet<>();
+        for (int numSequences = 0; numSequences < NUM_SEQUENCES && (refSeq = refSeqFile.nextSequence()) != null; ++numSequences) {
+            if (sequenceNames.contains(refSeq.getName())) {
+                throw new UserException.MalformedFile("Sequence name appears more than once in reference: " + refSeq.getName());
+            }
+            sequenceNames.add(refSeq.getName());
+            ret.add(makeSequenceRecord(refSeq));
+        }
+        return new SAMSequenceDictionary(ret);
+    }
+
+    /**
+     * Create one SAMSequenceRecord from a single fasta sequence
```

Comment:
let's either get rid of this code (by reading a dict directly) or at least document that it's copied from `CreateSequenceDictionary.makeSequenceRecord`